client/src/language/diagnosticsSupport.ts

```diff
@@ -64,14 +67,41 @@ export const updateDiagnostics = async (uri: vscode.Uri): Promise<void> => {
     }
     const newDiagnostic = {
       ...diagnostic,
-      range: newRange
+      range: newRange,
+      source: `${diagnostic.source}, ${diagnosticCollection.name}`
     }
     cleanDiagnostics.push(newDiagnostic)
   }))
-  const diagnosticCollection = diagnosticCollections[embeddedLanguageType]
   diagnosticCollection.set(originalTextDocument.uri, cleanDiagnostics)
 }
 
+// Review diagnostics that already exist for the documents of a recipe
+// This is intended to be called when a new scan finished, so diagnostics can be updated with the available information.
+export const reviewRecipeDiagnostics = async (recipeName: string): Promise<void> => {
```

Comment:
I think you can make this commit simpler. This commit is only about re-starting an analyzing when a recipe scan has completed. You probably have a function that can trigger a full embedded languages analyze of a document, right? A main analyze() function? I suggest simply calling it once a recipe scan has completed. And we already have the event for that, it's serverScanRequest. You don't need to add a new event.

Reply:
> The diagnostics are produced by the shell and python extensions as the user edit the documents. These extensions use their own internal logic to decide when to start the analyze. The only control we have on it is when we update the "embedded language documents", which we want to update as soon as possible in order to provide up to date completion items.
> 
> However, we have control over when we handle the diagnostics produces by these extensions. We could handle them only when the scan finished, but then old diagnostics will still be seen after the user has fixed them, and users won't see the new diagnostics for new errors they might have produced. In my opinion, it is best to update the diagnostics as soon as they are received (onDidChangeDiagnostics). The current implementation takes into account the result of the last scan when handling the new diagnostics produced by the Python and Shell extensions.
> 
> I'm thinking one alternative to reviewRecipeDiagnostics could be something like this:
> 
> ```ts
> export const reviewDiagnostics = async (): Promise<void> => {
>   const allDiagnostics = vscode.languages.getDiagnostics()
>   await Promise.all(allDiagnostics.map(async ([uri, _diagnostics]): Promise<void> => {
>     handleEmbeddedLanguageDocumentDiagnostics(uri) // to revert to previous name
>   }))
> }
> ```
> 
> It would be much easier to understand and maintain, but with the downside of doing unneeded work on diagnostics that do not need to be updated (which I don't expect to be a big deal). What do you think?
Seems right although I don't picture the full implementation yet.
> 
> About serverScanRequest, I tried to use it in a previous commit, but it brought a lot of overhead and I thought it would be better to use a simple "notify" function. I can't find the commit anymore, but I can find the commit where I removed it: https://github.com/yoctoproject/vscode-bitbake/compare/50077097a15a7d51e1eda92f25439ab660459e5d..2696cb114ef2e9a60b9e9c47a93f16b84521a1fb. Did I miss something?

What overhead did this add? That looks simpler than the current "notify" code IMO.

Reply:
done